src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -1335,21 +1255,24 @@ public Set<String> getDerivedPoolIdsForPools(Iterable<String> poolIds) {
      * @return
      *  A collection of entitlement IDs for the pools specified for the given pool IDs
      */
-    @SuppressWarnings("unchecked")
     public Collection<String> getEntitlementIdsForPools(Collection<String> poolIds) {
-        Set<String> eids = new HashSet<>();
+        Set<String> entitlementIds = new HashSet<>();
+
+        String jpql = """
+            SELECT DISTINCT e.id FROM Pool p
+            JOIN p.entitlements e
+            WHERE p.id IN :block
+            """;
 
         if (poolIds != null && !poolIds.isEmpty()) {
             for (List<String> block : this.partition(poolIds)) {
-                eids.addAll(this.currentSession().createCriteria(Pool.class, "p")
-                    .createAlias("p.entitlements", "e")
-                    .add(Restrictions.in("p.id", block))
-                    .setProjection(Projections.distinct(Projections.property("e.id")))
-                    .list());
+                entitlementIds.addAll(this.currentSession().createQuery(jpql, String.class)
```

Comment:
This is creating a new instance of the query for every iteration, which we should avoid. Additionally, this is going through the Hibernate session rather than the JPA entity manager to create the query. We should update this to use the latter.

Note this occurs in several places throughout this PR.

Reply:
done and we decide to split this task to 2 card to have smaller PRs for better review.